scripts-available/CDB_FederatedServerDiagnostics.sql

```diff
@@ -98,6 +98,86 @@ $$
 LANGUAGE SQL VOLATILE PARALLEL UNSAFE;
 
 
+--
+-- Get a foreign PG server hostname from the catalog
+--
+CREATE OR REPLACE FUNCTION @extschema@.__CDB_FS_Foreign_Server_Host_PG(server_internal name)
+RETURNS text
+AS $$
+    SELECT option_value FROM (
+        SELECT (pg_options_to_table(srvoptions)).*
+        FROM pg_foreign_server WHERE srvname = server_internal
+    ) AS opt WHERE opt.option_name = 'host';
+$$
+LANGUAGE SQL VOLATILE PARALLEL UNSAFE;
+
+
+--
+-- Get a foreign PG server port from the catalog
+--
+CREATE OR REPLACE FUNCTION @extschema@.__CDB_FS_Foreign_Server_Port_PG(server_internal name)
+RETURNS integer
+AS $$
+    SELECT option_value::integer FROM (
+        SELECT (pg_options_to_table(srvoptions)).*
+        FROM pg_foreign_server WHERE srvname = server_internal
+    ) AS opt WHERE opt.option_name = 'port';
+$$
+LANGUAGE SQL VOLATILE PARALLEL UNSAFE;
+
+--
+-- Get one measure of network latency in ms to a remote TCP server
+--
+CREATE OR REPLACE FUNCTION @extschema@.__CDB_FS_TCP_Network_Latency(
+    host text,
+    port integer,
+    timeout_seconds float DEFAULT 5.0
+)
+RETURNS float
+AS $$
+    import socket
+    from timeit import default_timer as timer
+
+    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    s.settimeout(timeout_seconds)
+
+    t_start = timer()
+
+    try:
+        s.connect((host, int(port)))
+        t_stop = timer()
+        s.shutdown(socket.SHUT_RD)
+    except (socket.timeout, OSError, socket.error):
+        plpy.warning('could not connect to server: %s:%d' % (host, port))
+        return None
+
+    t_total = (t_stop - t_start) * 1000.0
+    plpy.debug('TCP connection %s:%d time=%.2f ms' % (host, port, t_total))
+    return t_total
+$$
+LANGUAGE plpythonu VOLATILE PARALLEL UNSAFE;
+
+--
+-- Get the network latency to a remote PG server
+--
+CREATE OR REPLACE FUNCTION @extschema@.__CDB_FS_Foreign_Server_Latency_PG(server_internal name, n_samples integer DEFAULT 10)
+RETURNS float
+AS $$
+DECLARE
+    remote_server_host text := @extschema@.__CDB_FS_Foreign_Server_Host_PG(server_internal);
+    remote_server_port integer := @extschema@.__CDB_FS_Foreign_Server_Port_PG(server_internal);
+    latency float;
+BEGIN
+    latency := avg(sample) FROM (
```

Comment:
The average will hide errors. I'm inclined to say that we either want all the results or the max here.

Reply:
Then perhaps we can add in the same shot raw times, errors count, average, stdev and max, mimicking `ping`. The cost of doing so is marginal.